src/proxy.ts

```diff
@@ -0,0 +1,109 @@
+import { fork, Sequence } from 'effection';
+import { AddressInfo } from 'net';
+import * as proxy from 'http-proxy';
+import * as http from 'http';
+import { listen, ReadyCallback } from './http';
+import { EventEmitter, forkOnEvent, resumeOnEvent } from './util';
+import * as trumpet from 'trumpet';
+import * as zlib from 'zlib';
+
+interface ProxyOptions {
+  port: number,
+  targetPort: number
+  inject?: string
+};
+
+function pipe(from, to) {
+  fork(function*() {
+    let writer = forkOnEvent(from, 'data', function*(chunk) {
+      to.write(chunk);
+    });
+
+    try {
+      yield resumeOnEvent(from, 'end');
+      to.end();
+    } finally {
+      writer.halt();
+    }
+  });
+}
+
+export function* createProxyServer(options: ProxyOptions, ready: ReadyCallback = x => x): Sequence {
+  let proxyServer = proxy.createProxyServer({
+    target: `http://localhost:${options.targetPort}`,
+    selfHandleResponse: true
+  });
+
+  forkOnEvent(proxyServer, 'proxyRes', function*(proxyRes, req, res) {
+    for(let [key, value] of Object.entries(proxyRes.headers)) {
+      res.setHeader(key, value);
+    }
+
+    let contentType = proxyRes.headers['content-type'] as string;
+    let contentEncoding = proxyRes.headers['content-encoding'] as string;
+
+    if(contentType && contentType.split(';')[0] === 'text/html') {
+      res.removeHeader('content-length');
+      res.removeHeader('content-encoding');
+
+      res.writeHead(proxyRes.statusCode, proxyRes.statusMessage);
+
+      let tr = trumpet();
+
+      tr.select('head', (node) => {
+        let rs = node.createReadStream();
+        let ws = node.createWriteStream();
+        ws.write(options.inject || '');
+        rs.pipe(ws);
+      });
+
+      if(contentEncoding && contentEncoding.toLowerCase() == 'gzip') {
+        let zip = zlib.createGunzip();
+
+        pipe(proxyRes, zip);
+        pipe(zip, tr);
+      } else {
+        pipe(proxyRes, tr);
+      }
+
+      pipe(tr, res);
```

Comment:
Ideally, we want to `yield` here, and wait for the request to finish. That seems like it'd look clearer, and also gives us the option of cleaning up once we're finished (e.g. cleaning up the callback handler above). I'm not sure how best to structure the code to this though.